### PR TITLE
msg/async: deallocate protocol in EventCenter.

### PR DESCRIPTION
**Pull request by rzarzynski**
Helgrind complains about:

  ```
  ==00:00:04:55.495 510301== Possible data race during read of size 8 at 0x1EF1B0B0 by thread #96
  ==00:00:04:55.495 510301== Locks held: 3, at addresses 0xEE8D650 0xEE8D748 0x1FFEFFEE08
  ==00:00:04:55.495 510301==    at 0x107F5C5: ~unique_ptr (unique_ptr.h:273)
  ==00:00:04:55.495 510301==    by 0x107F5C5: ~AuthConnectionMeta (Auth.h:163)
  ==00:00:04:55.495 510301==    by 0x107F5C5: std::_Sp_counted_ptr<AuthConnectionMeta*, (__gnu_cxx::_Lock_policy)2>::_M_dispose() (shared_ptr_base.h:377)
  ==00:00:04:55.495 510301==    by 0x10C19EE: _M_release (shared_ptr_base.h:155)
  ==00:00:04:55.495 510301==    by 0x10C19EE: _M_release (shared_ptr_base.h:148)
  ==00:00:04:55.495 510301==    by 0x10C19EE: ~__shared_count (shared_ptr_base.h:728)
  ==00:00:04:55.495 510301==    by 0x10C19EE: ~__shared_ptr (shared_ptr_base.h:1167)
  ==00:00:04:55.495 510301==    by 0x10C19EE: ~shared_ptr (shared_ptr.h:103)
  ==00:00:04:55.495 510301==    by 0x10C19EE: Protocol::~Protocol() (Protocol.cc:14)
  ==00:00:04:55.495 510301==    by 0x1081A5C: ProtocolV2::~ProtocolV2() (ProtocolV2.cc:100)
  ==00:00:04:55.495 510301==    by 0x105AA48: operator() (unique_ptr.h:81)
  ==00:00:04:55.495 510301==    by 0x105AA48: ~unique_ptr (unique_ptr.h:274)
  ==00:00:04:55.495 510301==    by 0x105AA48: AsyncConnection::~AsyncConnection() (AsyncConnection.cc:149)
  ==00:00:04:55.495 510301==    by 0x105ABAC: AsyncConnection::~AsyncConnection() (AsyncConnection.cc:154)
  ==00:00:04:55.495 510301==    by 0xD2BAA0: RefCountedObject::put() const (RefCountedObj.cc:27)
  ==00:00:04:55.495 510301==    by 0xEB1E58: intrusive_ptr_release (RefCountedObj.h:193)
  ==00:00:04:55.495 510301==    by 0xEB1E58: ~intrusive_ptr (intrusive_ptr.hpp:98)
  ==00:00:04:55.495 510301==    by 0xEB1E58: ~pair (stl_pair.h:208)
  ==00:00:04:55.495 510301==    by 0xEB1E58: destroy<std::pair<const entity_addrvec_t, boost::intrusive_ptr<AsyncConnection> > > (new_allocator.h:140)
  ==00:00:04:55.495 510301==    by 0xEB1E58: destroy<std::pair<const entity_addrvec_t, boost::intrusive_ptr<AsyncConnection> > > (alloc_traits.h:487)
  ==00:00:04:55.495 510301==    by 0xEB1E58: _M_deallocate_node (hashtable_policy.h:2100)
  ==00:00:04:55.495 510301==    by 0xEB1E58: _M_erase (hashtable.h:1909)
  ==00:00:04:55.495 510301==    by 0xEB1E58: std::_Hashtable<entity_addrvec_t, std::pair<entity_addrvec_t const, boost::intrusive_ptr<AsyncConnection> >, std::allocator<std::pair<entity_addrvec_t const, boost
  ::intrusive_ptr<AsyncConnection> > >, std::__detail::_Select1st, std::equal_to<entity_addrvec_t>, std::hash<entity_addrvec_t>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__
  detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::erase(std::__detail::_Node_const_iterator<std::pair<entity_addrvec_t const, boost::intrusive_ptr<AsyncConnection> >, fals
  e, true>) (hashtable.h:1884)
  ==00:00:04:55.495 510301==    by 0xEB24EA: erase (hashtable.h:762)
  ==00:00:04:55.495 510301==    by 0xEB24EA: erase (unordered_map.h:798)
  ==00:00:04:55.495 510301==    by 0xEB24EA: AsyncMessenger::_lookup_conn(entity_addrvec_t const&) (AsyncMessenger.h:312)
  ==00:00:04:55.495 510301==    by 0xEAA08A: AsyncMessenger::connect_to(int, entity_addrvec_t const&, bool, bool) (AsyncMessenger.cc:701)
  ==00:00:04:55.495 510301==    by 0xEF7477: connect_to_mon (Messenger.h:530)
  ==00:00:04:55.495 510301==    by 0xEF7477: MonClient::_add_conn(unsigned int, unsigned long) (MonClient.cc:716)
  ==00:00:04:55.495 510301==    by 0xEF7C30: MonClient::_add_conns(unsigned long) (MonClient.cc:779)
  ==00:00:04:55.495 510301==    by 0xEFDED6: MonClient::_reopen_session(int) (MonClient.cc:691)
  ==00:00:04:55.495 510301==    by 0xF025F2: MonClient::tick() (MonClient.cc:925)
  ==00:00:04:55.495 510301==    by 0x72A06C: Context::complete(int) (Context.h:77)
  ==00:00:04:55.495 510301==    by 0xD35276: SafeTimer::timer_thread() (Timer.cc:96)
  ==00:00:04:55.495 510301==    by 0xD36850: SafeTimerThread::entry() (Timer.cc:30)
  ==00:00:04:55.495 510301==    by 0xA7E1203: mythread_wrapper (hg_intercepts.c:389)
  ==00:00:04:55.495 510301==    by 0xC6182DD: start_thread (in /usr/lib64/libpthread-2.28.so)
  ==00:00:04:55.495 510301==    by 0xD8B34B2: clone (in /usr/lib64/libc-2.28.so)
  ==00:00:04:55.495 510301==  Address 0x1ef1b0b0 is 128 bytes inside a block of size 136 alloc'd
  ==00:00:04:55.495 510301==    at 0xA7DC0C3: operator new[](unsigned long) (vg_replace_malloc.c:433)
  ==00:00:04:55.495 510301==    by 0x1082094: ProtocolV2::reset_recv_state()::{lambda()#2}::operator()() const [clone .isra.916] (ProtocolV2.cc:240)
  ==00:00:04:55.495 510301==    by 0x108244B: EventCenter::C_submit_event<ProtocolV2::reset_recv_state()::{lambda()#2}>::do_request(unsigned long) (Event.h:228)
  ==00:00:04:55.495 510301==    by 0xEB7465: EventCenter::process_events(unsigned int, std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> >*) (Event.cc:433)
  ==00:00:04:55.495 510301==    by 0xEBC1BB: operator() (Stack.cc:53)
  ==00:00:04:55.495 510301==    by 0xEBC1BB: std::_Function_handler<void (), NetworkStack::add_thread(unsigned int)::{lambda()#1}>::_M_invoke(std::_Any_data const&) (std_function.h:297)
  ==00:00:04:55.495 510301==    by 0xCF4AA32: ??? (in /usr/lib64/libstdc++.so.6.0.25)
  ==00:00:04:55.495 510301==    by 0xA7E1203: mythread_wrapper (hg_intercepts.c:389)
  ==00:00:04:55.495 510301==    by 0xC6182DD: start_thread (in /usr/lib64/libpthread-2.28.so)
  ==00:00:04:55.495 510301==    by 0xD8B34B2: clone (in /usr/lib64/libc-2.28.so)
  ==00:00:04:55.495 510301==  Block was alloc'd by thread #3
  ```

Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/msg/async/AsyncConnection.cc [hunk 1]
```diff
@@ -151,6 +151,11 @@ AsyncConnection::~AsyncConnection()
   if (recv_buf)
     delete[] recv_buf;
   ceph_assert(!delay_state);
+  // protocol must be deallocated in proper EventCenter. This is ensured
```
Comment (tchaikov) [inline src/msg/async/AsyncConnection.cc]: what about `recv_buf`?

Comment (rzarzynski) [reply src/msg/async/AsyncConnection.cc]: Apparently, it isn't touched outside the EC. Constructor only allocates the memory.

File: src/msg/async/AsyncConnection.cc [hunk 2]
```diff
@@ -744,6 +749,7 @@ void AsyncConnection::cleanup() {
     delete delay_state;
     delay_state = NULL;
   }
+  protocol.reset();
```
Comment (tchaikov) [inline src/msg/async/AsyncConnection.cc]:
```
2020-02-13T12:03:54.142 INFO:tasks.ceph.mon.a.smithi102.stderr:*** Caught signal (Segmentation fault) **
2020-02-13T12:03:54.142 INFO:tasks.ceph.mon.a.smithi102.stderr: in thread 7fa128cdf700 thread_name:fn_monstore
2020-02-13T12:03:54.143 INFO:tasks.ceph.mgr.x.smithi107.stderr:2020-02-13T12:03:54.141+0000 7faa7be28040 -1 WARNING: all dangerous and experimental features are enabled.
2020-02-13T12:03:54.143 INFO:tasks.ceph.mgr.y.smithi102.stderr:*** Caught signal (Segmentation fault) **
2020-02-13T12:03:54.143 INFO:tasks.ceph.mgr.y.smithi102.stderr: in thread 7faae27ef040 thread_name:ceph-mgr
2020-02-13T12:03:54.145 INFO:tasks.ceph.mgr.x.smithi107.stderr:2020-02-13T12:03:54.142+0000 7faa7be28040 -1 WARNING: all dangerous and experimental features are enabled.
2020-02-13T12:03:54.145 INFO:tasks.ceph.mon.a.smithi102.stderr: ceph version 15.1.0-596-g3be473c (3be473c7d578040af701705444ac5bec35ae4f4a) octopus (rc)
2020-02-13T12:03:54.145 INFO:tasks.ceph.mon.a.smithi102.stderr: 1: (()+0x12dc0) [0x7fa1358b7dc0]
2020-02-13T12:03:54.145 INFO:tasks.ceph.mon.a.smithi102.stderr: 2: (AsyncConnection::is_connected()+0xb) [0x7fa137e11f2b]
2020-02-13T12:03:54.145 INFO:tasks.ceph.mon.a.smithi102.stderr: 3: (PaxosService::dispatch(boost::intrusive_ptr<MonOpRequest>)+0xda) [0x5597fedad61a]
2020-02-13T12:03:54.145 INFO:tasks.ceph.mon.a.smithi102.stderr: 4: (PaxosService::C_RetryMessage::_finish(int)+0x64) [0x5597fed17d94]
2020-02-13T12:03:54.145 INFO:tasks.ceph.mon.a.smithi102.stderr: 5: (C_MonOp::finish(int)+0x49) [0x5597fecbbf79]
2020-02-13T12:03:54.146 INFO:tasks.ceph.mon.a.smithi102.stderr: 6: (Context::complete(int)+0xd) [0x5597fecb933d]
2020-02-13T12:03:54.146 INFO:tasks.ceph.mon.a.smithi102.stderr: 7: (void finish_contexts<std::__cxx11::list<Context*, std::allocator<Context*> > >(CephContext*, std::__cxx11::list<Context*, std::allocator<Context*> >&, int)+0xa5) [0x5597fece15d5]
2020-02-13T12:03:54.146 INFO:tasks.ceph.mon.a.smithi102.stderr: 8: (PaxosService::_active()+0x1cb) [0x5597fedaeebb]
2020-02-13T12:03:54.146 INFO:tasks.ceph.mon.a.smithi102.stderr: 9: (Context::complete(int)+0xd) [0x5597fecb933d]
2020-02-13T12:03:54.146 INFO:tasks.ceph.mon.a.smithi102.stderr: 10: (void finish_contexts<std::__cxx11::list<Context*, std::allocator<Context*> > >(CephContext*, std::__cxx11::list<Context*, std::allocator<Context*> >&, int)+0xa5) [0x5597fece15d5]
2020-02-13T12:03:54.146 INFO:tasks.ceph.mon.a.smithi102.stderr: 11: (Paxos::finish_round()+0x169) [0x5597feda49d9]
2020-02-13T12:03:54.146 INFO:tasks.ceph.mon.a.smithi102.stderr: 12: (Paxos::commit_finish()+0x91c) [0x5597feda755c]
2020-02-13T12:03:54.147 INFO:tasks.ceph.mon.a.smithi102.stderr: 13: (C_Committed::finish(int)+0x45) [0x5597fedab8f5]
2020-02-13T12:03:54.147 INFO:tasks.ceph.mon.a.smithi102.stderr: 14: (Context::complete(int)+0xd) [0x5597fecb933d]
2020-02-13T12:03:54.147 INFO:tasks.ceph.mon.a.smithi102.stderr: 15: (MonitorDBStore::C_DoTransaction::finish(int)+0x98) [0x5597fedab618]
2020-02-13T12:03:54.147 INFO:tasks.ceph.mon.a.smithi102.stderr: 16: (Context::complete(int)+0xd) [0x5597fecb933d]
2020-02-13T12:03:54.147 INFO:tasks.ceph.mon.a.smithi102.stderr: 17: (Finisher::finisher_thread_entry()+0x1a5) [0x7fa137bfb5c5]
2020-02-13T12:03:54.147 INFO:tasks.ceph.mon.a.smithi102.stderr: 18: (()+0x82de) [0x7fa1358ad2de]
2020-02-13T12:03:54.147 INFO:tasks.ceph.mon.a.smithi102.stderr: 19: (clone()+0x43) [0x7fa134657133]
2020-02-13T12:03:54.147 INFO:tasks.ceph.mon.a.smithi102.stderr:2020-02-13T12:03:54.142+0000 7fa128cdf700 -1 *** Caught signal (Segmentation fault) **
2020-02-13T12:03:54.148 INFO:tasks.ceph.mon.a.smithi102.stderr: in thread 7fa128cdf700 thread_name:fn_monstore
```

see http://pulpito.ceph.com/kchai-2020-02-13_11:45:07-rados-wip-kefu-testing-2020-02-13-0025-distro-basic-smithi/

Comment (rzarzynski) [reply src/msg/async/AsyncConnection.cc]: Oops, yet another race. Apparently somebody calls this after `cleanup()`.

```cpp
bool AsyncConnection::is_connected() {
  return protocol->is_connected();
}
```

**Pull request closed without merging**